### PR TITLE
Async mongodb driver

### PR DESCRIPTION
**Pull request by allanbank**
This pull request contains a new driver for MongoDB based on the [asynchronous driver](http://github.com/allanbank/mongodb-async-driver). The bulk of the changes are adding the required logic to the startup scripts to support the second driver. I have used this branch to run a series of [benchmarks](http://www.allanbank.com/mongodb-async-driver/performance/ycsb.html).

There are also a number of community contributed updates to the old MongoDB Driver. These include:

* Updated README with specific instructions for running a MongoDB suite.
* Updates to the configuration to now be strictly based on a [MongoDB Connection String URI](http://docs.mongodb.org/manual/reference/connection-string/) which is standardized across drivers.
* Other small fixes and performance enhancements.

This pull request will supersede the following pull requests:

* [PR-112](https://github.com/brianfrankcooper/YCSB/pull/112) - Integrates the MongoDB Connection String URI for both drivers.
* [PR-115](https://github.com/brianfrankcooper/YCSB/pull/115) - The writeConcern and readPreference are expressed via the standardized  MongoDB Connection String URI.
* [PR-131](https://github.com/brianfrankcooper/YCSB/pull/131) - Read preferences are passed via the MongoDB Connection String and replica sets are automatically detected for the asynchronous driver and will be enabled with the MongoDB Inc. driver by passing multiple servers in the URI.

I know this is a large volume of changes but they are all focused on the MongoDB driver and thought it would be easier to provide them in one batch than a bunch of pull requests. If you would prefer a bunch of smaller changes I will need guidance on the desired granularity.

### CODE REVIEW

**Comment by busbey** [comment]
I restarted the build now that the mapkeeper stuff is out of the way. I'd like to get this set of changes merged in before cutting the next release.

If someone from the mongo community can work with me on needed updates, I'll proactively go point the related open issues here and close them.

File: bin/ycsb [hunk 1]
```diff
@@ -107,9 +108,15 @@ command = COMMANDS[sys.argv[1]]["command"]
 database = sys.argv[2]
 db_classname = DATABASES[database]
 options = sys.argv[3:]
+java_home = os.environ["JAVA_HOME"]
```
Comment (busbey) [inline bin/ycsb]: the java home stuff is good, but seems unrelated to the rest of the mongo improvements. If I file an issue for "allow the use of java home" could you edit the commit message to call out closing it?

File: mongodb/README.md [hunk 2]
```diff
@@ -1,44 +1,95 @@
...
-    cd mongodb-linux-x86_64-1.8.3
+    cd mongodb-linux-x86_64-*
     ./bin/mongod --dbpath /tmp/mongodb
 
-### 2. Set Up YCSB
+Replace x.x.x above with the latest stable release version for MongoDB.
+See http://docs.mongodb.org/manual/installation/ for installation steps for various operating systems.
 
-Clone the YCSB git repository and compile:
+### 2. Install Java and Maven
```
Comment (busbey) [inline mongodb/README.md]: This sounds like a gap in our overall YCSB docs. individual db bindings shouldn't have to tell folks to go get java.

I'll write a follow on issue to deal with this across modules.

File: mongodb/src/main/java/com/yahoo/ycsb/db/AsyncMongoDbClient.java [hunk 3]
```diff
@@ -0,0 +1,477 @@
+/**                                                                                                                                                                                
+ * Copyright (c) 2013-2014, Allanbank Consulting, Inc. All rights reserved.
```
Comment (busbey) [inline mongodb/src/main/java/com/yahoo/ycsb/db/AsyncMongoDbClient.java]: This will take some time for me to review how our licensing ought to work. Generally, for the contributions I've already seen to the project, the practice is for contributors to assign copyrights to Yahoo! (as the project originator) and then use an ASL license.

File: mongodb/src/main/java/com/yahoo/ycsb/db/AsyncMongoDbClient.java [hunk 4]
```diff
@@ -0,0 +1,477 @@
...
+import com.yahoo.ycsb.DBException;
+
+/**
+ * MongoDB asynchronous client for the YCSB framework.
+ * <p>
+ * See the README for details on configuration the client and driver.
+ * </p>
+ * 
+ * @author rjm
+ * @copyright 2013-2014, Allanbank Consulting, Inc., All Rights Reserved
```
Comment (busbey) [inline mongodb/src/main/java/com/yahoo/ycsb/db/AsyncMongoDbClient.java]: the project relies on license headers at the top of files and not copyright javadocs.

File: mongodb/src/main/java/com/yahoo/ycsb/db/AsyncMongoDbClient.java [hunk 5]
```diff
@@ -0,0 +1,477 @@
...
+                        (BinaryElement) be));
+            }
+        }
+    }
+
+    /**
+     * BinaryByteArrayIterator provides an adapter from a {@link BinaryElement}
+     * to a {@link ByteIterator}.
+     * 
+     * @copyright 2013, Allanbank Consulting, Inc., All Rights Reserved
```
Comment (busbey) [inline mongodb/src/main/java/com/yahoo/ycsb/db/AsyncMongoDbClient.java]: definitely no need for another copyright notice down here.

File: mongodb/src/main/java/com/yahoo/ycsb/db/MongoDbClient.java [hunk 6]
```diff
@@ -72,52 +75,41 @@ public void init() throws DBException {
 
             // initialize MongoDb driver
             Properties props = getProperties();
+
+            // Just use the standard connection format URL
+            // http://docs.mongodb.org/manual/reference/connection-string/
             String url = props.getProperty("mongodb.url",
-                    "mongodb://localhost:27017");
-            database = props.getProperty("mongodb.database", "ycsb");
-            String writeConcernType = props.getProperty("mongodb.writeConcern",
```
Comment (busbey) [inline mongodb/src/main/java/com/yahoo/ycsb/db/MongoDbClient.java]: since we're moving the write concern from a property to a connection URL, these changes will have to go into a new major version of YCSB since people with existing test rigs for running MongoDB will break.

That's fine, just a heads up about timing.

File: pom.xml [hunk 7]
```diff
@@ -50,7 +50,8 @@
     <infinispan.version>7.1.0.CR1</infinispan.version>
     <openjpa.jdbc.version>2.1.1</openjpa.jdbc.version>
     <mapkeeper.version>1.0</mapkeeper.version>
-    <mongodb.version>2.11.2</mongodb.version>
+    <mongodb.version>2.12.2</mongodb.version>
```
Comment (busbey) [inline pom.xml]: can we update this to a more recent version? I know #221 was trying to go to 2.13.0

**Comment by allanbank** [comment]
@busbey - Give me a day or two to make the changes you suggested.  Is it OK if I rebase the pull on the latest master making the changes as I go?

File: bin/ycsb (hunk 1 above)
Comment (allanbank) [inline bin/ycsb]: Will do just let me know the issue number or if you want me to write the issue.

File: mongodb/src/main/java/com/yahoo/ycsb/db/AsyncMongoDbClient.java (hunk 3 above)
Comment (allanbank) [inline mongodb/src/main/java/com/yahoo/ycsb/db/AsyncMongoDbClient.java]: I will update the license.  Is there a assignment document that need to be submitted?

File: mongodb/src/main/java/com/yahoo/ycsb/db/MongoDbClient.java (hunk 6 above)
Comment (allanbank) [inline mongodb/src/main/java/com/yahoo/ycsb/db/MongoDbClient.java]: I will make it backwards compatible and note that the URL supersedes the old properties.

File: pom.xml (hunk 7 above)
Comment (allanbank) [inline pom.xml]: I will update to the latest version.

**Comment by busbey** [comment]
please feel free to rebase. I prefer rebases to lots of merges, personally. :)

File: pom.xml (hunk 7 above)
Comment (busbey) [inline pom.xml]: just a heads up, I'd like to merge #253 while this stuff is in flight since it's so small.

File: mongodb/src/main/java/com/yahoo/ycsb/db/MongoDbClient.java (hunk 6 above)
Comment (busbey) [inline mongodb/src/main/java/com/yahoo/ycsb/db/MongoDbClient.java]: perfect!

File: pom.xml (hunk 7 above)
Comment (allanbank) [inline pom.xml]: OK - let me know when that is merged done.

Comment (busbey) [inline pom.xml]: done. :)

File: mongodb/src/main/java/com/yahoo/ycsb/db/AsyncMongoDbClient.java (hunk 3 above)
Comment (busbey) [inline mongodb/src/main/java/com/yahoo/ycsb/db/AsyncMongoDbClient.java]: let me touch base with other maintainers. I can't find anything in the project docs.

File: bin/ycsb (hunk 1 above)
Comment (busbey) [inline bin/ycsb]: great. presuming you go forward with the rebase, please edit the commit message for this part of the changes to note "fixes #252"

**Comment by allanbank** [comment]
@busbey -

I think that latest push covers all of the issues you raised. Let me know if I missed something.

Rob.

**Comment by busbey** [comment]
The update looks great.

It looks like with the current update we'll be using the Mongo 3 client, right? Will that prevent us from talking to Mongo 2 servers? If so, how comon are Mongo 2 installations? Is it worth maintaining both, ala our HBase binding?

**Comment by allanbank** [comment]
The driver can talk to version 2 and 3 clusters.

Rob

On May 23, 2015 12:22:12 PM EDT, Sean Busbey <notifications@github.com> wrote:
>The update looks great.
>
>It looks like with the current update we'll be using the Mongo 3
>client, right? Will that prevent us from talking to Mongo 2 servers? If
>so, how comon are Mongo 2 installations? Is it worth maintaining both,
>ala our HBase binding?
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/brianfrankcooper/YCSB/pull/180#issuecomment-104920362

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by busbey** [comment]
Thanks for sticking with this contribution! I hope to see more great stuff from the mongo community in the future.

**Comment by allanbank** [comment]
@busbey - No problem. Thanks for doing the merge. 

I mentioned it in the initial pull request but I am pretty certain these pull requests can be closed now.

* [PR-112](https://github.com/brianfrankcooper/YCSB/pull/112) - Integrates the MongoDB Connection String URI for both drivers.
* [PR-115](https://github.com/brianfrankcooper/YCSB/pull/115) - The writeConcern and readPreference are expressed via the standardized  MongoDB Connection String URI.
* [PR-131](https://github.com/brianfrankcooper/YCSB/pull/131) - Read preferences are passed via the MongoDB Connection String and replica sets are automatically detected for the asynchronous driver and will be enabled with the MongoDB Inc. driver by passing multiple servers in the URI.

If you would like me to add an explicit comment to each of those explaining why it can be closed I'll be happy to do that.

Rob.

**Comment by busbey** [comment]
Thanks for the reminder, I'll go close them out.
